src/w25m.rs

```diff
@@ -0,0 +1,154 @@
+pub struct Die0;
+#[allow(missing_debug_implementations)]
+pub struct Die1;
+
+/// All dies which are supposed to be supported by the W25M struct have to implement this trait
+pub trait Stackable<SPI: Transfer<u8>, CS: OutputPin>: BlockDevice<SPI, CS> + Read<SPI, CS> + Sized {
+    fn new(spi: SPI, cs: CS) -> Result<Self, Error<SPI, CS>>;
+    /// Returns the SPI and chip select objects so they can be used elsewhere
+    fn free(self) -> (SPI, CS);
+}
+
+#[derive(Debug)]
+pub struct Flash<DIE0, DIE1, DIE> 
+{
+    inner: Inner<DIE0, DIE1>,
+    _die: PhantomData<DIE>,
+}
+
+#[derive(Debug)]
+enum Inner<DIE0, DIE1> {
+    Die0(DIE0),
+    Die1(DIE1),
+    Dummy,
+}
+
+impl<DIE0, DIE1> Flash<DIE0, DIE1, Die0> 
+{
+    /// Creates a new W25M device
+    /// 
+    /// At 
+    /// the moment the only way to call this function is sadly
+    /// ```
+    /// let mut flash: Flash<W25N<_, _>, W25N<_, _>, _> = Flash::init(spi, cs).unwrap();
+    /// ```
+    /// TODO: Improve this API, its not very convenient
+    pub fn init<SPI, CS>(spi: SPI, cs: CS) -> Result<Flash<DIE0, DIE1, Die0>, Error<SPI, CS>> 
+        where 
+        SPI: Transfer<u8>,
+        CS: OutputPin,
+        DIE0: Stackable<SPI, CS>,
+        DIE1: Stackable<SPI, CS>,
+    {
+        Ok(Flash{
+            inner: Inner::Die0(DIE0::new(spi, cs)?),
+            _die: PhantomData
+        })
+    }
+}
+
+impl<DIE0, DIE1> Flash<DIE0, DIE1, Die0> 
+
+{
+    pub fn switch_die<SPI, CS>(mut self) -> Result<Flash<DIE0, DIE1, Die1>, Error<SPI, CS>>
+    where DIE0: Stackable<SPI, CS>,
+    DIE1: Stackable<SPI, CS>,
+    SPI: Transfer<u8>,
+    CS: OutputPin {
+        let (mut spi, mut cs) = match mem::replace(&mut self.inner, Inner::Dummy) {
+            Inner::Die0(die) => die.free(),
+            _ => unreachable!()
+        };
+        let mut command = [0xC2, 0x01];
+        cs.set_low().map_err(Error::Gpio)?;
+        let spi_result = spi.transfer(&mut command).map_err(Error::Spi);
+        cs.set_high().map_err(Error::Gpio)?;
+        spi_result?;
+
+        Ok(Flash{
+            inner: Inner::Die1(DIE1::new(spi, cs)?),
+            _die: PhantomData
+        })
+    }
+}
+
+impl<DIE0, DIE1> Flash<DIE0, DIE1, Die1> 
+{
+    pub fn switch_die<SPI, CS>(mut self) -> Result<Flash<DIE0, DIE1, Die0>, Error<SPI, CS>> 
+        where DIE0: Stackable<SPI, CS>,
+    DIE1: Stackable<SPI, CS>,
+    SPI: Transfer<u8>,
+    CS: OutputPin 
+    {
+        let (mut spi, mut cs) = match mem::replace(&mut self.inner, Inner::Dummy) {
+            Inner::Die1(die) => die.free(),
+            _ => unreachable!()
+        };
+
+        let mut command = [0xC2, 0x00];
+        cs.set_low().map_err(Error::Gpio)?;
+        let spi_result = spi.transfer(&mut command).map_err(Error::Spi);
+        cs.set_high().map_err(Error::Gpio)?;
+        spi_result?;
+
+
+        Ok(Flash{
+            inner: Inner::Die0(DIE0::new(spi, cs)?),
+            _die: PhantomData
+        })
+    }
+}
```

Comment:
The `switch_die` functions should call into a common function to reduce the code duplication